### ISSUE TITLE
useFetcher cannot infer type of data if data is an array of objects with `data` key

### ISSUE THREAD

**Issue by wolthers** [remix-run/remix#5607]
### What version of Remix are you using?

1.13.0

### Are all your remix dependencies & dev-dependencies using the same version?

- [X] Yes

### Steps to Reproduce

```tsx
import { useFetcher } from "@remix-run/react";

export default function Component () {
  type MyWronglyInferredResponse = Array<{
    data: {
      key1: number;
      key2: string
    }
  }>
  const wrongFetcher = useFetcher<MyWronglyInferredResponse>()

  if (wrongFetcher.state === "idle" && wrongFetcher.type === "done") {
    wrongFetcher.data
    // ---------- ⬆️
    // (property) data: SerializeDeferred<{
    //     key1: number;
    //     key2: string;
    // }>[]
  }

  type MyCorrectlyInferredResponse = Array<{
    someOtherKeyThanData: {
      key1: number;
      key2: string
    }
  }>
  const correctFetcher = useFetcher<MyCorrectlyInferredResponse>()

  if (correctFetcher.state === "idle" && correctFetcher.type === "done") {
    correctFetcher.data
    // ------------ ⬆️
    // (property) data: SerializeObject<UndefinedToOptional<{
    //   someOtherKeyThanData: {
    //       key1: number;
    //       key2: string;
    //   };
    // }>>[]
  }
}
```

### Expected Behavior

wrongFetcher.data should work as correctFetcher.

### Actual Behavior

wrongFetcher isn't inferring the returned data correctly.

I'm almost certain that this change by @jacob-ebey introduced the bug in 1.11.0: https://github.com/remix-run/remix/pull/4920/files#diff-73d5f419b6a2e6f026b68dd66fcf6ac019072255943a18bf69fa94acc2451746R23

**Comment by machour** [comment]
Probably related to #5211

**Comment by machour** [comment]
This will be fixed in 1.16.0 🎉 
Here's the result with 1.16.0-pre.0:

<img width="907" alt="Screenshot 2023-04-29 at 1 20 48 PM" src="https://user-images.githubusercontent.com/304450/235302266-1c780291-4960-4563-b402-4f51d923b681.png">
